### PR TITLE
[#11147] Student view profile: preserve proportions of the profile image

### LINKED ISSUE #11147: Student view profile: preserve proportions of the profile image
Context: a student is viewing the profile photo

Current:
image

Problem: the image is squashed into a fixed size and with. We can't assume the image to have a specific width:height ratio because no such requirement is mention or enforced at the when uploading the image.

Suggestion: Fix only the width (or height), preserving proportions of the image

### PR DESCRIPTION
**Pull request by drblaui**
Fixes #11147 

**Outline of Solution**
The problem of the cause was the css file belonging to the Student Profile, where pics are forced into 220px x 220px. This was fixed by capping the width and height at a maximum of 220px with `max-height` and `max-width` and letting CSS calculate the image size automatically. 

Two tests were also added that test an image that's taller than wider and an image that's wider than taller

### CODE REVIEW

**Comment by drblaui** [comment]
@wkurniawan07 I think these tests fail because of #11462 hasn't been merged and selenium isn't able to upload pictures, since these tests work on my machine. Can you or anyone on your team look into this?

File: src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java [hunk 1]
```diff
@@ -42,6 +42,27 @@ public void testAll() {
...
+        ______TS("Typical case: Profile picture ratios");
+        profilePage.fillProfilePic("src/test/resources/images/profile_pic_too_wide.jpg");
+        profilePage.uploadPicture();
+        profilePage.verifyStatusMessage("Your profile picture has been saved successfully");
+
+        profilePage.showPictureEditor(); // In order to upload
+        profilePage.waitForUploadEditModalVisible();
+
+        profilePage.editProfilePhoto();
+        profilePage.verifyPhotoMaxHeight(220, "src/test/resources/images/profile_pic_too_wide.jpg");
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java:54]: I don't find this necessary. The test pictures used are already known beforehand and thus the dimensions can be readily determined. Just use the same method `verifyPhotoSize` with the correct values.

File: src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java [hunk 2]
```diff
@@ -42,6 +42,27 @@ public void testAll() {
...
 
+        ______TS("Typical case: Profile picture ratios");
+        profilePage.fillProfilePic("src/test/resources/images/profile_pic_too_wide.jpg");
+        profilePage.uploadPicture();
+        profilePage.verifyStatusMessage("Your profile picture has been saved successfully");
+
+        profilePage.showPictureEditor(); // In order to upload
+        profilePage.waitForUploadEditModalVisible();
+
+        profilePage.editProfilePhoto();
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java:53]: This shouldn't be necessary as this block's intention is not to test the edit action

File: src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java [hunk 3]
```diff
@@ -42,6 +42,27 @@ public void testAll() {
         profilePage.editProfilePhoto();
         profilePage.verifyPhotoSize("220px", "220px");
 
+        ______TS("Typical case: Profile picture ratios");
+        profilePage.fillProfilePic("src/test/resources/images/profile_pic_too_wide.jpg");
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java:46]: The lines from `profilePage.fillProfilePic` to `profilePage.waitForUploadEditModalVisible` are repeatable 3 times. They should be abstracted out.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Here I am assuming the local file storage problem (which should be fixed now) is not the cause of the test failure

**Comment by drblaui** [comment]
I will look into this probably in the middle of next week, thanks for the input!

File: src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java (hunk 2 above)
Comment (drblaui) [reply src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java:53]: Your're right, will remove this

File: src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java (hunk 3 above)
Comment (drblaui) [reply src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java:46]: I understand, will also be changed

File: src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java (hunk 2 above)
Comment (drblaui) [reply src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java:53]: After some interesting experience, I've found out that (for whatever reason) `profilePage.showPictureEditor();` is necessary for the tests to succeed. I have no idea why, I can in real time see that after selenium uploads the profile pictures, they are scaled right but for whatever reason the tests still think they are 220px by 220px. The other two lines are however as you said useless

**Comment by drblaui** [comment]
@wkurniawan07 is everything okay now?

File: src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java [hunk 4]
```diff
@@ -159,9 +159,13 @@ public void showPictureEditor() {
...
     }
 
-    public void verifyPhotoSize(String height, String width) {
-        assertEquals(height, browser.driver.findElement(By.className("profile-pic")).getCssValue("height"));
-        assertEquals(width, browser.driver.findElement(By.className("profile-pic")).getCssValue("width"));
+    public void verifyPhotoSize(int height, int width) {
+        String browserHeight = browser.driver.findElement(By.className("profile-pic")).getCssValue("height");
+        float imageHeight = Float.parseFloat(browserHeight.substring(0, browserHeight.length() - 2));
+        String browserWidth = browser.driver.findElement(By.className("profile-pic")).getCssValue("width");
+        float imageWidth = Float.parseFloat(browserWidth.substring(0, browserWidth.length() - 2));
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java:166]: `browserWidth.substring(0, browserWidth.length() - 2)` is a magic operation and `2` in it is a magic number. I know the intention here, not necessarily everyone else. At the very least where `2` comes from should be made clearer.

File: src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java [hunk 5]
```diff
@@ -40,7 +40,11 @@ public void testAll() {
         profilePage.waitForUploadEditModalVisible();
 
         profilePage.editProfilePhoto();
-        profilePage.verifyPhotoSize("220px", "220px");
+        profilePage.verifyPhotoSize(220, 220);
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java:43]: This block can also reuse the `uploadAndVerify` function. `editProfilePhoto` is a separate test operation that can be done after checking the photo size.

File: src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java [hunk 6]
```diff
@@ -50,4 +54,12 @@ public void testAll() {
         profilePage.ensureProfileContains("short.name", "e@email.tmt", "inst", "American",
                 StudentProfileAttributes.Gender.FEMALE, "this is enough!$%&*</>");
     }
+
+    private void uploadAndVerify(StudentProfilePage profilePage, String image, int height, int width) {
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java:58]: Upload and verify what exactly?

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java [hunk 7]
```diff
@@ -50,4 +50,12 @@ public void testAll() {
         profilePage.ensureProfileContains("short.name", "e@email.tmt", "inst", "American",
                 StudentProfileAttributes.Gender.FEMALE, "this is enough!$%&*</>");
     }
+
+    private void uploadProfilePicAndVerifyDimensions(StudentProfilePage profilePage, String image, int height, int width) {
+        profilePage.fillProfilePic(image);
+        profilePage.uploadPicture();
+        profilePage.verifyStatusMessage("Your profile picture has been saved successfully");
+        profilePage.showPictureEditor(); // In order to upload
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java:58]: No need for the comment `// In order to upload`

File: src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java [hunk 8]
```diff
@@ -159,12 +159,24 @@ public void showPictureEditor() {
...
+    public void closePictureEditor() {
         click(uploadEditModal.findElement(By.className("close")));
     }
 
+    public void verifyPhotoSize(int height, int width) {
+        String browserHeight = browser.driver.findElement(By.className("profile-pic")).getCssValue("height");
+        String browserWidth = browser.driver.findElement(By.className("profile-pic")).getCssValue("width");
+        /*
+         * Remove the last two chars from the string, so we can isolate the Measurement from
+         * "[MEASUREMENT]px", since we want to parse it into a float
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java:171]: Having this comment doesn't change the fact that `2` is still a magic number. Why not do something more self-explanatory like:
```java
browserHeight.replaceFirst("px$", "")
```

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java (hunk 8 above)
Comment (drblaui) [reply src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java:171]: Because I actually didn't know that replaceFirst() exists, thanks

**Review by daongochieu2810 — APPROVED** [review]

File: src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java [hunk 9]
```diff
@@ -50,4 +50,13 @@ public void testAll() {
         profilePage.ensureProfileContains("short.name", "e@email.tmt", "inst", "American",
                 StudentProfileAttributes.Gender.FEMALE, "this is enough!$%&*</>");
     }
+
+    private void uploadProfilePicAndVerifyDimensions(StudentProfilePage profilePage, String imagePath,
+                                                     int height, int width) {
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/cases/StudentProfilePageE2ETest.java:55]: This should me a method in `StudentProfilePage`, not the E2E test class

File: src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java [hunk 10]
```diff
@@ -159,12 +159,21 @@ public void showPictureEditor() {
...
 
+    public void verifyPhotoSize(int height, int width) {
+        String browserHeight = browser.driver.findElement(By.className("profile-pic")).getCssValue("height");
+        String browserWidth = browser.driver.findElement(By.className("profile-pic")).getCssValue("width");
+        //Remove px from Dimension Strings to turn it into a number
+        float imageHeight = Float.parseFloat(browserHeight.replaceFirst("px$", ""));
+        float imageWidth = Float.parseFloat(browserWidth.replaceFirst("px$", ""));
+        assertEquals(height, imageHeight, 1.0);
+        assertEquals(width, imageWidth, 1.0);
+        closePictureEditor();
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java:174]: A "verify" action should not have any business with closing picture editor. If you need this to be done on all operations, it should be part of the `uploadProfilePicAndVerifyDimensions` method instead.

File: src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java [hunk 11]
```diff
@@ -159,12 +159,21 @@ public void showPictureEditor() {
...
-        assertEquals(height, browser.driver.findElement(By.className("profile-pic")).getCssValue("height"));
-        assertEquals(width, browser.driver.findElement(By.className("profile-pic")).getCssValue("width"));
+    public void closePictureEditor() {
         click(uploadEditModal.findElement(By.className("close")));
     }
 
+    public void verifyPhotoSize(int height, int width) {
+        String browserHeight = browser.driver.findElement(By.className("profile-pic")).getCssValue("height");
+        String browserWidth = browser.driver.findElement(By.className("profile-pic")).getCssValue("width");
+        //Remove px from Dimension Strings to turn it into a number
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java:169]: This is not necessary; the code reads exactly what this comment says (i.e. comment doesn't value-add). That's the advantage of using good abstractions.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

**Comment by drblaui** [comment]
@wkurniawan07 the InstructorAuditLogsPageE2ETest seems to fail, but I never touched that file and as far as I can see my changes have nothing to do with that test. Do you have any insights?

File: src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java [hunk 12]
```diff
@@ -159,12 +159,29 @@ public void showPictureEditor() {
...
+        String browserHeight = browser.driver.findElement(By.className("profile-pic")).getCssValue("height");
+        String browserWidth = browser.driver.findElement(By.className("profile-pic")).getCssValue("width");
+        float imageHeight = Float.parseFloat(browserHeight.replaceFirst("px$", ""));
+        float imageWidth = Float.parseFloat(browserWidth.replaceFirst("px$", ""));
+        assertEquals(height, imageHeight, 1.0);
+        assertEquals(width, imageWidth, 1.0);
+    }
+
+    public void uploadProfilePicAndVerifyDimensions(String imagePath,
+                                                     int height, int width) {
```
Comment (wkurniawan07) [inline src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java:176]: This should be on the previous line. Looks kind of weird to have line break when it obviously can fit one line.

**Review by wkurniawan07 — APPROVED** [review]
LGTM, just one piece of nitpicking

File: src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java (hunk 12 above)
Comment (drblaui) [reply src/e2e/java/teammates/e2e/pageobjects/StudentProfilePage.java]: Oh yeah that seems to be remaining from when I still passed the pageobject into it and lint complained. Should be prettier now

**Pull request merged**